roc.go

```diff
@@ -81,41 +89,39 @@ func ROC(n int, y []float64, classes []bool, weights []float64) (tpr, fpr []floa
 		}
 		nPos += posWeight
 		nNeg += negWeight
-		tpr[bin] += posWeight
-		fpr[bin] += negWeight
-
-		// Assess if the bin needs to be updated. If n is zero,
-		// the bin is always updated, unless consecutive y values
-		// are equal. Otherwise, the bin must be updated until it
-		// matches the next y value (skipping empty bins).
-		if n == 0 {
-			if i != (len(y)-1) && y[i] != y[i+1] {
-				bin++
-				tpr[bin] = tpr[bin-1]
-				fpr[bin] = fpr[bin-1]
-			}
-		} else {
-			for i != (len(y)-1) && y[i+1] > tol {
-				tol += incWidth
-				bin++
-				tpr[bin] = tpr[bin-1]
-				fpr[bin] = fpr[bin-1]
-			}
+		if y[i] <= cutoffs[bin] {
+			tpr[bin] += posWeight
+			fpr[bin] += negWeight
 		}
 	}
-	if n == 0 {
-		tpr = tpr[:(bin + 1)]
-		fpr = fpr[:(bin + 1)]
-	}
 
 	invNeg := 1 / nNeg
 	invPos := 1 / nPos
 	for i := range tpr {
 		tpr[i] *= invPos
 		fpr[i] *= invNeg
 	}
-	tpr[len(tpr)-1] = 1
-	fpr[len(fpr)-1] = 1
 
 	return tpr, fpr
 }
+
+// EquallySpaced returns n equally spaced values,
```

Comment:
Yep, I'll just use Span instead. I knew there had to be a function somewhere that did that but I just didn't know where it was, thanks! As for why cutoffs[0] has to be slightly less than y[0] -- this is so that the ROC curve will start at (0,0) -- as in the description, y values less than or equal to the cutoff are classified as true. If the equality was in the other direction, you would need the last cutoff to be slightly greater than the maximum in order for the ROC curve to end at (1,1).